### PR TITLE
Re-organize board-specific samples per vendor and add zephyr:code-sample directive 

### PR DESCRIPTION
**Pull request by kartben**
- Added missing `.. zephyr:code-sample::` directives
- Grouped samples by vendor
- Added intermediate table of contents to get rid of the very flat/messy top level ToC we had

Note 1: The size of the change set is impressive, but it really is mostly just moving files around while adding proper HTML redirects when necessary. Splitting this into multiple PRs would lead to mostly systematic merge conflicts in the `redirects.py` file.

Note 2: `samples/boards/arc_secure_services` probably needs a better home as it's not clear it's a "board" sample. I have left it out from the PR and this is the only remaining "non-vendor" top-level folder underneath samples/boards.

BEFORE: https://docs.zephyrproject.org/latest/samples/boards/boards.html
AFTER: https://builds.zephyrproject.io/zephyr/pr/78089/docs/samples/boards/boards.html

### CODE REVIEW

**Review by erwango — DISMISSED** [review]

**Review by butok — DISMISSED** [review]
Thanks for cleaning!

**Comment by kartben** [comment]
> @kartben Thanks for this valuable clean up. From what I understand, samples are now sorted by `zephyr:code-sample::`  and appears using `name`

@erwango 

Right now, nothing has changed in terms of sorting order, i.e. it's based on the name of the folders. In the future, yes, it's the `:name:` that will by default be picked up for sorting alphabetically. You may check https://github.com/zephyrproject-rtos/zephyr/pull/77400 for a sneak peek ; I think you will like it :)

> so if we want to do further clean up (like sorting and more explicit titles) we should play with those, am I correct ?

Short answer: yes, names should definitely be "consistent" so that samples that belong next to one another effectively appear as such. 
Long(er) answer: I will look into adding a way to tweak the ordering, e.g. by means of indicating "featured" samples in a given category (see more about "categories" below). But I don't want to overthink/over-engineer it, and would really like the default avenue to be sticking to short, very descriptive names, that group reasonably well by "just" sorting alphabetically.

> If we want to do more sorting (along with the increase of those vendor specific samples), will it be possible to add sub sections (like bluetooth, power mgmt, ...). Note that I didn't care much previously, but since it now appears as nice usable pages, we can make more usage of those (like providing the link in documentation and so on).

Yes :) The PR I just linked will be introducing *categories*, for further grouping. To your point of vendor specific samples starting to get crowded, the idea is that it would first rely on logical grouping in the repo (i.e, you'd want to group the power mgmt samples in boards/stm32/pm), and then the new category directive allows to provide a short description (and human readable name, at a minimum) of said category.
See e.g.:
- https://github.com/kartben/zephyr/blob/samples_toc/samples/drivers/drivers.rst
- https://github.com/kartben/zephyr/blob/samples_toc/samples/drivers/adc/index.rst
- https://builds.zephyrproject.io/zephyr/pr/77400/docs/samples/drivers/adc/index.html

**Comment by kartben** [comment]
> will it be possible to add sub sections (like bluetooth, power mgmt, ...). Note that I didn't care much previously, but since it now appears as nice usable pages, we can make more usage of those (like providing the link in documentation and so on).

@erwango  oh well, I forgot I had actually started to do just this in the linked PR 🤣 
https://builds.zephyrproject.io/zephyr/pr/77400/docs/samples/boards/st/index.html

File: samples/boards/stm32/power_mgmt/wkup_pins/README.rst [hunk 1]
```diff
@@ -1,7 +1,8 @@
-.. _gpio-as-a-wkup-pin-src-sample:
+.. zephyr:code-sample:: stm32_pm_gpio_wkup_src
+   :name: GPIO as a wake-up pin source
+   :relevant-api: adc_interface
```
[the remark this answers is not in the document]
Comment (kartben) [reply samples/boards/stm32/power_mgmt/wkup_pins/README.rst]: thanks -- updated to `:relevant-api: sys_poweroff gpio_interface`

File: samples/boards/st/bluetooth/interactive_gui/README.rst [hunk 2]
```diff
@@ -30,7 +30,7 @@ The UART default settings are:
 Building and Running
 ********************
 
-This sample can be found under :zephyr_file:`samples/boards/stm32/bluetooth/interactive_gui` in the
+This sample can be found under :zephyr_file:`samples/boards/stm/bluetooth/interactive_gui` in the
```
[the remark this answers is not in the document]
Comment (kartben) [reply samples/boards/st/bluetooth/interactive_gui/README.rst]: thanks!

**Review by gmarull — DISMISSED** [review]

**Review by erwango — DISMISSED** [review]

**Comment by mmahadevan108** [comment]
@kartben, please rebase

**Comment by kartben** [comment]
@mmahadevan108 thanks for spotting this needed to be rebased :) done!

**Review by JarmouniA — APPROVED** [review]

**Review by nordicjm — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Comment by marekmatej** [comment]
Espressif parts LGTM!  Thanks for this helpful update!

**Review by marekmatej — APPROVED** [review]

**Comment by kartben** [comment]
This has had approvals (either in current version of the PR on earlier version before a rebase was needed) from all the major vendors impacted by the change. I will be re-assigning to myself as this is in the samples area at the end of the day, and to make sure this can be merged sooner than later as it is likely to generate merge conflicts. Thanks everyone for the reviews!

**Pull request merged**